### PR TITLE
[Arm64] Implement managed Short Vector ABI for Vector64<T> Vector128<T>

### LINKED ISSUE #14371: [Arm64] Support for Vector Intrinsics and/or SIMD requires 128bit argument registers to be preserved
@dotnet/jit-contrib @dotnet/arm64-contrib 

The following logic and assumptions will no longer be true when SIMD intrinsic support is added.
``` C++
//--------------------------------------------------------------------
// This represents the floating point argument registers which are saved
// as part of the NegInfo for a FramedMethodFrame. Note that these
// might not be saved by all stubs: typically only those that call into
// C++ helpers will need to preserve the values in these volatile
// registers.
//--------------------------------------------------------------------
typedef DPTR(struct FloatArgumentRegisters) PTR_FloatArgumentRegisters;
struct FloatArgumentRegisters {
    // armV8 supports 32 floating point registers. Each register is 128bits long.
    // It can be accessed as 128-bit value or 64-bit value(d0-d31) or as 32-bit value (s0-s31)
    // or as 16-bit value or as 8-bit values. C# only has two builtin floating datatypes float(32-bit) and 
    // double(64-bit). It does not have a quad-precision floating point.So therefore it does not make sense to
    // store full 128-bit values in Frame when the upper 64 bit will not contain any values.
    double  d[8];  // d0-d7
};
```

I plan to change this to save ful 128-bits all the time.  If there are any objections, let me know.

This will also affect `SAVE_FLOAT_ARGUMENT_REGISTERS` and `RESTORE_FLOAT_ARGUMENT_REGISTERS` macros and related assembly frame layouts

category:correctness
theme:vector-codegen
skill-level:expert
cost:small

### LINKED ISSUE #16021: [Arm64] Add Short Vector ABI support for HW Intrinsics
The types `Vector64<T>` and `Vector128<T>` are intended to map to ABI `Short Vector` fundamental types.  These should be passed in vector registers.

Current code will pass these in integer registers.

`IsShortVector` can be crudely implemented as follows:
```C#
bool Compiler::IsShortVector(CORINFO_CLASS_HANDLE hClass)
{
    return isIntrinsicType(hClass) && (getBaseTypeOfSIMDType(hClass) != TYP_UNKNOWN);
}
```

`Compiler::getPrimitiveTypeForStruct()` can be trivially updated to return `TYP_SIMD8` or `TYP_SIMD16` when `IsShortVector() == true`

I made an initial naive attempt to implement and failed, but I really need a design discussion.

@CarolEidt @tannergooding @dotnet/arm64-contrib @dotnet/jit-contrib 

category:reach
theme:intrinsics
skill-level:expert
cost:medium

### LINKED ISSUE #16022: [Arm64] Add HVA ABI support for HW Intrinsics
Once Short Vector ABI is supported #16021, Homogenous Short Vector Aggregates (HVA) should be supported.

One approach would be to treat these as a subclass of `HFA`, `getHfaType()` could then return `TYP_SIMD8` or `TYP_SIMD16`.  This may simplify implementation.

@CarolEidt @tannergooding @dotnet/arm64-contrib @dotnet/jit-contrib This will require a design discussion

category:reach
theme:intrinsics
skill-level:expert
cost:medium

### PR DESCRIPTION
**Pull request by sdmaclea**
This change implements changes necessary to support `Vector64<T>` `Vector128<T>` as Homogeneous Vector Aggregate (`HVA`) types each containing a single `Short Vector`.

It should also implement most of the change necessary to also handle multiple element `HVA`  in managed code.  I actually believe the functional implementation is complete.  Lacking test coverage it is difficult to confirm this also works.  I will write a test shortly.

This change is fairly extensive.  I would appreciate an early review.

Current status:
+ All Arm64 tests are passing with this change except one test is failing in `LinearScan::checkLastUses()` (debugging now) 
+ Need comments regarding use of `#ifdef`.  
    + I am tempted to add a `FEATURE_ARM64_HVA_ABI`
    + I added `#ifdef` in most places even when it felt like the code should be common
+ Need comments related to VM to JIT interface
   + I probably need to change the GUID
+ Need comments relative to whether this should be done.  
  + For Arm64 if we plan to do this, now would be good.  I have spent 6 days writing and debugging this so far.  I would hate to scrap it and then have to redo the work in 6 months.
+ Need comments about skipping treating types as a struct containing a Short Vector.  
  + It feels natural to me, but it was not the original plan

Fixes #14371
Fixes #16021
Prepares to fix #16022

@janvorli @jkotas @RussKeldorph @CarolEidt @tannergooding PTAL
@dotnet/jit-contrib @dotnet/arm64-contrib FYI

### CODE REVIEW

**Comment by tannergooding** [comment]
Is it "accurate" to treat `short vectors` as HFA?

My understanding of the spec is that `short vectors` are a "fundamental data type" and that you could construct an HFA using them, but that they themselves are not HFAs.

Additionally, I had thought that any type which is composed of short vectors would actually make it a `HVA`, which is similar (but not identical) to a HFA.

**Comment by sdmaclea** [comment]
> My understanding of the spec is that short vectors are a "fundamental data type" and that you could construct an HFA/HVA using them, but that they themselves are not HFA/HVAs.

Correct.

> Additionally, I had thought that any type which is composed of short vectors would generally make it a HVA, which is similar (but not identical) to a HFA.

Correct.

To be semantically correct, I should change all references to HFA to HomogeneousAggregate (or any other generic name).  

It didn't seem necessary or worth it, but it could easily be done with a global search and replace.  If it was renamed, it should be done in a separate PR.

SVE will introduce two new vector types and two new aggregate types.
+ Scalable Vector, Scalable Predicates, Scalable Vector Aggregates, Scalable Predicate Aggregates

It seemed silly to treat these all differently, when the only difference would be the aggregated type.

> Is it "accurate" to treat short vectors as HFA?

It is accurate to treat `Vector64<T>` and `Vector128<T>` as an HVA containing one Short Vector.

It also could be accurate to treat `Vector64<T>` and `Vector128<T>` as an Short Vector.  While this was a  viable option, this was less convenient and added an extra use case which didn't need to be supported in managed code.

In the JIT we treat a single element homogenous aggregate as the primitive type (`TYP_DOUBLE`,`TYP_FLOAT`, `TYP_SIMD16`, or `TYP_SIMD8`) for ABI purposes.  So from a JIT perspective it is only a semantic difference.

Observation:
+ `float`, `double`, 8 & 16 byte `short vectors` are all passed in the same registers using the same conventions as each other.
+ A struct containing `float`, `double`, 8 & 16 byte `short vectors`   are all passed in the same registers using the same conventions as the bare primitives.
+ Similarly a struct containing 2-4 `float`, `double`, `short vectors`   are all passed in the same registers using the same conventions as each other.
+ Similarly a struct containing multiple nested structs containing `float`, `double`, `short vectors` are all passed in the same registers using the same conventions as the unnested structs.

This is all logical because the Homogenous Aggregates rules are based on in memory layout.  Struct is a higher level language construct which doesn't affect the ABI.

**Comment by tannergooding** [comment]
Looking at the spec, I think the primary benefit in differentiating is that there are a number of steps which differ slightly for a fundamental type vs a HVA/HFA type that can ultimately make the argument passing logic simpler for a given type.

That being said, I think that can probably be tracked by a work item and fixed/handled later (But I do think it will be worth fixing, eventually).

**Comment by sdmaclea** [comment]
> I think the primary benefit in differentiating is that there are a number of steps which differ slightly for a fundamental type vs a HVA/HFA type that can ultimately make the argument passing logic simpler for a given type.

`Stage C – Assignment of arguments to registers and stack`

Can you be more specific?  I do not see a functional difference.  JIT already treated `C.1` as equivalent to `C.2` in `getPrimitiveTypeForStruct()`

**Comment by sdmaclea** [comment]
OK, maybe `already` was a stretch.  Since `Vector64<T>` and `Vector128<T>` are structs it seemed already to me.

**Comment by tannergooding** [comment]
> Can you be more specific?

`Stage C` has 15 steps. Several of those apply only to integer fundamental types, several apply only to floating-point fundamental types, and several only apply to HVA/HFA.

While it does look to be the case that an HVA containing a single `short vector` ends up being treated the same as if you just treated it as `short vector` directly, the HFA/HVA specific steps end up doing several things that it is possible to short-circuit for `short vector` types directly (such as alignment checks and size adjustments). The same goes for integer fundamental types in comparison with composite types.

File: src/inc/corhdr.h [hunk 1]
```diff
@@ -913,6 +913,9 @@ typedef enum CorElementType
     ELEMENT_TYPE_SENTINEL       = 0x01 | ELEMENT_TYPE_MODIFIER, // sentinel for varargs
     ELEMENT_TYPE_PINNED         = 0x05 | ELEMENT_TYPE_MODIFIER,
 
+    ELEMENT_TYPE_SHORT_VECTOR   = 0x80,
```
Comment (jkotas) [inline src/inc/corhdr.h]: This file is meant to contain only the standard `ELEMENT_TYPE_` constants that show up in regular metadata.

The various custom area specific should rather be somewhere else. corpriv.h may be a good place - there are several of them already (look for `ELEMENT_TYPE_`).

File: src/vm/class.cpp [hunk 2]
```diff
@@ -1664,6 +1664,31 @@ CorElementType MethodTable::GetHFAType()
         _ASSERTE(pMT->IsValueType());
```
Comment (jkotas) [inline src/vm/class.cpp]: Do you need to modify all callers of GetHFAType to handle these new values correctly?

Comment (sdmaclea) [reply src/vm/class.cpp]: I thought I looked at them all,but maybe I maybe I waiting for an assert to fire...  I'll take a look next week.

**Comment by sdmaclea** [comment]
@tannergooding 
> such as alignment checks and size adjustments

For Arm64 C#, these are treated the same as `Short Vector`, since 
+ the Short Vector size is always a multiple of 8 bytes
+ structs (and therefore Short Vector) are always 8 byte aligned
+ HVA are 1-4 Short Vectors (no layout)

I looked at those operations as `noop` (git spelling) or freebies

I can believe I missed something.  When we look at native interoperability, we may find something important.

**Comment by sdmaclea** [comment]
@dotnet-bot help

**Comment by sdmaclea** [comment]
Fixed broken test, rebased, fixed merge conflicts, and enabled simd tests now that Arm64 HW intrinsics are included in corefx packages.

**Comment by sdmaclea** [comment]
This PR got a bit crufty while I was working on enabling official arm64 builds.

This is important for arm64 HW intrinsics ZBB.  I will work on fixing this.

**Comment by sdmaclea** [comment]
@dotnet-bot help

**Pull request closed without merging**